sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java

```diff
@@ -480,7 +484,13 @@ Mono<Void> closeAsync(AmqpShutdownSignal shutdownSignal) {
                     .log("Closed CBS node.")),
             managementNodeCloseOperations.doFinally(signalType ->
                 logger.atVerbose()
+                    .addKeyValue(SIGNAL_TYPE_KEY, signalType)
                     .log("Closed management nodes.")))
+            // Make sure to close request-response-channel before session is closed.
+            .then(emitShutDownSignalOperation.doFinally(signalType ->
+                logger.atVerbose()
```

Comment:
One case I could think of is - With the connection waiting for closure of RRChannel before notifying all the ReactorReceiver(s), a potential delay (default timeout: 60sec) in RRChannel's close (e.g., due to remote-close didn't complete) further delays recovery of ReactorReceiver(s), which means downstream wait for more time while recovery is happening. 

Today, the new connection gets created immediately, serving downstream, while the old connection closure happens in parallel.

Reply:
Thanks for telling me about the different between Explicit closeAysnc and Implict closeAysnc.

It is worthwhile to think about various scenarios. I took lots of time to think about my changes as it does change the existing closing order. 

I do this change for two reasons:
1. After adding `.takeUntilOther(shutdownSignalSink.asMono())`, I have to make sure the CBS channel closed earlier than shutdown signal emitted.
2. When we emit shutdown signal, the ReactorSession may closed ealier than CBS channel, so the Close frame is sent earlier than Detach frame, which cause the CBS channel timeout. The connection may always closed after 60sec timeout.  This change can solve this issue.

For the case you mentioned, right, the change will postpone the closing of ReactorRecevier(s). But I have a question, so for each connection, they can have its own CBS channel, and old CBS channel closure has no impact to the creation of new one. But they are using the same ReactorReceiver, so "old" ReactoreRecevier closure will block "new" ReactorReceiver, right? I will check the Implicit closeAsync, this part of logic seems very interesting.

Reply:
> they are using the same ReactorReceiver, so "old" ReactorRecevier closure will block "new" ReactorReceiver, right?

I read the RCA in #29212 and got the answer, the ReactorReceiver(s) are control by AmqpReceiveLinkProcessors, if old "ReactorReceiver" haven't closed, the new one won't be created. 

As the closing timeout will be added for ReactorReceiver in PR #29201, I am afraid if ReactorReceiver also has the timeout issue when it closed (Detach frame be sent later than Close frame). 

So I am thinking about a way to emit the shutdown ReactorSession and RectorSender/ReactorReceiver signal separately. Then we can close RectorSender/ReactorReceiver earlier.

Reply:
> I read the RCA in https://github.com/Azure/azure-sdk-for-java/issues/29212 and got the answer, the ReactorReceiver(s) are control by AmqpReceiveLinkProcessors, if old "ReactorReceiver" haven't closed, the new one won't be created.

Yep, that's right.

>  I am afraid if ReactorReceiver also has the timeout issue when it closed (Detach frame be sent later than Close frame).

if we look at ReactorSession, we can see multiple routes it gets disposed/closed. In all routes, except one, the ReactorSession will send CLOSE only after child ReactorReceiver(s) closes, meaning it gives the ReactorReceiver(s) a chance to send DETACH. The only one route where ReactorSession sends CLOSE without waiting for ReactorReaciver(s) DETACH is when the Connection sends the shutdown signal to the ReactorSession. I think this is an optimization, given doing so will be considered as implicit disposal of links under the session on the broker side (less traffic). 

I'm working on a pr to have no-wait time in this particular case, but it will still make sure to terminate the endpoint state (the issue we fixed in [29201](https://github.com/Azure/azure-sdk-for-java/pull/29201)); hence AmqpReceiveLinkProcessor gets to react.

Reply:
That makes sense. I will rethink my changes. I could get some thoughts from your changes for ReactorReceiver. Maybe treat differently for explicit close and implicit close.

Reply:
Sounds good; I'll open that pr and tag you to get thoughts; I'm trying to achieve it without adding extra aync edge and nodes (to reduce the async signal passing+thread hoping) to the underlying graph. The more the async-signaling, the more it's hard to debug/fix in the future.

Reply:
I move `emitShutdownSignalOperation` into the `Mono.whenDelayError`, so ReactorReceiver(s) do need to wait cbs node completely closed. I have done several explict cases test and works fine and trying to repro some recovery cases.

>The only one route where ReactorSession sends CLOSE without waiting for ReactorReaciver(s) DETACH is when the Connection sends the shutdown signal to the ReactorSession.

I reconsidered this scenario, because the closing of ReactorReceiver(s) does not prevent the connection from closing (unlike the closing of a CBS node which prevents the connection from closing), ReactorReceiver(s) can also be closed after the connection is closed. So, I think having no-wait time close for ReactorReceiver is nice-to-have as it can close receiver earlier.
I also done some tests, for example, explicitly closing EH consumer/processor, throw exception to close Connection, all these scenarios will close ReactorReceiver(s) first and then close Connection. For both logic analysis and tests, that timeout issue occurs in CBS node may not occur in ReactorReceiver(s). I'll keep checking the logic of connection closing.